lib/services/local/generic/common.go

```diff
@@ -0,0 +1,59 @@
+// Teleport
+// Copyright (C) 2024 Gravitational, Inc.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package generic
+
+import (
+	"context"
+	"time"
+
+	"github.com/gravitational/teleport/lib/backend"
+	"github.com/gravitational/teleport/lib/services"
+)
+
+// MarshalFunc is a type signature for a marshaling function.
+type MarshalFunc[T any] func(T, ...services.MarshalOption) ([]byte, error)
+
+// UnmarshalFunc is a type signature for an unmarshalling function.
+type UnmarshalFunc[T any] func([]byte, ...services.MarshalOption) (T, error)
+
+// ServiceCommon is a common generic service interface.
+type ServiceCommon[T any] interface {
+	// WithPrefix will return a service with the given parts appended to the backend prefix.
+	WithPrefix(parts ...string) ServiceCommon[T]
+	// GetResources returns a list of all resources.
+	GetResources(ctx context.Context) ([]T, error)
+	// ListResources returns a paginated list of resources.
+	ListResources(ctx context.Context, pageSize int, pageToken string) ([]T, string, error)
+	// GetResource returns the specified resource.
+	GetResource(ctx context.Context, name string) (resource T, err error)
+	// CreateResource creates a new resource.
+	CreateResource(ctx context.Context, resource T) (T, error)
+	// UpdateResource updates an existing resource.
+	UpdateResource(ctx context.Context, resource T) (T, error)
+	// UpsertResource upserts a resource.
+	UpsertResource(ctx context.Context, resource T) (T, error)
+	// DeleteResource removes the specified resource.
+	DeleteResource(ctx context.Context, name string) error
+	// DeleteAllResources removes all resources.
+	DeleteAllResources(ctx context.Context) error
+	// UpdateAndSwapResource will get the resource from the backend, modify it, and swap the new value into the backend.
+	UpdateAndSwapResource(ctx context.Context, name string, modify func(T) error) (T, error)
+	// MakeBackendItem will check and make the backend item.
+	MakeBackendItem(resource T, name string) (backend.Item, error)
```

Comment:
`MakeBackendItem(resource T, name string) (backend.Item, error)` is a low level "construct an object" operation, which doesn't touch backend. There is a single external user for this function, which makes me wonder if this usage is valid or not. The service in question `SAMLIdPServiceProviderService` uses it like so (the other call site is very similar):
```go
	item, err := s.svc.MakeBackendItem(sp, sp.GetName())
	if err != nil {
		return trace.Wrap(err)
	}

	return trace.Wrap(s.svc.RunWhileLocked(ctx, samlIDPServiceProviderModifyLock, samlIDPServiceProviderModifyLockTTL,
		func(ctx context.Context, backend backend.Backend) error {
			if err := s.ensureEntityIDIsUnique(ctx, sp); err != nil {
				return trace.Wrap(err)
			}

			_, err := backend.Update(ctx, item)
			if trace.IsNotFound(err) {
				return trace.NotFound("%s %q doesn't exist", types.KindSAMLIdPServiceProvider, sp.GetName())
			}

			return trace.Wrap(err)
		}))
```

Reply:
> which doesn't touch backend

Could we mention that in the godoc?

> There is a single external user for this function

That makes me feel it could be dropped. Maybe leave a TODO if you are so inclined?